commit 3019062905533ccd71b535f6e5c3ed7e9bb195e0
Author: Abdiel Janulgue <abdiel.janulgue@linux.intel.com>
Date:   Fri Dec 15 12:20:55 2017 +0200

    drm/i915: Ignore TMDS clock limit for DP++ when EDID override is set
    
    4K modes testing by using dummy EDID data has never been working
    properly on boxes with DP++ (dual-mode) adaptors. The reason for
    this is that those modes got pruned during hdmi mode validation.
    intel_hdmi_mode_valid returns CLOCK_HIGH because the pixel clock
    reported by the 4k mode is higher than dual port TMDS clock limit.
    
    However 4k injection does work properly on machines that don't have
    DP++ adapters because the mode is never validated against the DP++
    TMDS clock limit.
    
    v2: Don't detect the DP++ limits when we're testing using overridden
        EDIDs. Make sure to check for the override condition after
        respecting the value of drm_dp_dual_mode_detect (Jani Nikula).
    
    Bugzilla: https://bugs.freedesktop.org/show_bug.cgi?id=101649
    Cc: Ville Syrjälä <ville.syrjala@linux.intel.com>
    Cc: Jani Nikula <jani.nikula@intel.com>
    Cc: Daniel Vetter <daniel.vetter@ffwll.ch>
    Signed-off-by: Abdiel Janulgue <abdiel.janulgue@linux.intel.com>
    Signed-off-by: Jani Nikula <jani.nikula@intel.com>
    Link: https://patchwork.freedesktop.org/patch/msgid/20171215102055.11729-1-abdiel.janulgue@linux.intel.com

diff --git a/drivers/gpu/drm/i915/intel_hdmi.c b/drivers/gpu/drm/i915/intel_hdmi.c
index 179d0ad3889d..93e97f9e1aac 100644
--- a/drivers/gpu/drm/i915/intel_hdmi.c
+++ b/drivers/gpu/drm/i915/intel_hdmi.c
@@ -1567,7 +1567,10 @@ intel_hdmi_dp_dual_mode_detect(struct drm_connector *connector, bool has_edid)
 	 * there's nothing connected to the port.
 	 */
 	if (type == DRM_DP_DUAL_MODE_UNKNOWN) {
-		if (has_edid &&
+		/* An overridden EDID imply that we want this port for testing.
+		 * Make sure not to set limits for that port.
+		 */
+		if (has_edid && !connector->override_edid &&
 		    intel_bios_is_port_dp_dual_mode(dev_priv, port)) {
 			DRM_DEBUG_KMS("Assuming DP dual mode adaptor presence based on VBT\n");
 			type = DRM_DP_DUAL_MODE_TYPE1_DVI;